stream_processing/streaming_to_datalake.py

```diff
@@ -0,0 +1,158 @@
+                        .config("spark.sql.execution.arrow.pyspark.enabled", "true")
+                        .appName("Streaming Processing Application")
+                        .getOrCreate()
+        )
+        
+        logging.info('Spark session successfully created!')
+
+    except Exception as e:
+        traceback.print_exc(file=sys.stderr)
+        logging.error(f"Couldn't create the spark session due to exception: {e}")
+
+    return spark
+
+
+def create_initial_dataframe(spark_session):
+    """
+        Reads the streaming data and creates the initial dataframe accordingly
+    """
+    try: 
+        df = (spark_session
+            .readStream
+            .format("kafka")
+            .option("kafka.bootstrap.servers", "localhost:9092")
+            .option("subscribe", "device.iot.taxi_nyc_time_series")
+            # .option("startingOffsets", "earliest")
+            .option("failOnDataLoss", "false")
+            .load())
+        logging.info("Initial dataframe created successfully!")
+    except Exception as e:
+        logging.warning(f"Initial dataframe could not be created due to exception: {e}")
+
+    return df
+
+
+def create_final_dataframe(df, spark_session):
+    """
+        Modifies the initial dataframe, and creates the final dataframe
+    """
+    from pyspark.sql.types import IntegerType, StringType, StructType, StructField, TimestampNTZType, DoubleType, LongType
+    from pyspark.sql.functions import col, from_json, udf
+
+    payload_after_schema = StructType([
+            StructField("dolocationid", IntegerType(), True),
+            StructField("pulocationid", IntegerType(), True),
+            StructField("ratecodeid", DoubleType(), True),
+            StructField("vendorid", IntegerType(), True),
+            StructField("congestion_surcharge", DoubleType(), True),
+            StructField("extra", DoubleType(), True),
+            StructField("fare_amount", DoubleType(), True),
+            StructField("improvement_surcharge", DoubleType(), True),
+            StructField("mta_tax", DoubleType(), True),
+            StructField("passenger_count", DoubleType(), True),
+            StructField("payment_type", IntegerType(), True),
+            StructField("tip_amount", DoubleType(), True),
+            StructField("tolls_amount", DoubleType(), True),
+            StructField("total_amount", DoubleType(), True),
+            StructField("tpep_dropoff_datetime", LongType(), True),
+            StructField("tpep_pickup_datetime", LongType(), True),
+            StructField("trip_distance", DoubleType(), True)
+    ])
+
+    data_schema = StructType([
+        StructField("payload", StructType([
+            StructField("after", payload_after_schema, True)
+        ]), True)
+    ])
+
+    parsed_df = df.selectExpr("CAST(value AS STRING) as json") \
+                .select(from_json(col("json"), data_schema).alias("data")) \
+                .select("data.payload.after.*")
+
+    parsed_df = parsed_df \
```

Comment:
Explain this code!

Reply:
Explain:
(1) df.selectExpr("CAST(value AS STRING) as json")
--> Converts the value column of df to a STRING type and names the new column 'json'.
(2) .select(from_json(col("json"), data_schema).alias("data"))
--> Converts the new 'json' column into JSON format based on the schema 'data_schema' with the alias 'data'.
(3) .select("data.payload.after.*")
--> Based on the Debezium JSON, only extracts the data from payload.after.*

Reply:
=)) Explain in the code =)) not explain to me